src/boot/bootctl-install.c

```diff
@@ -215,7 +217,32 @@ static int version_check(int fd_from, const char *from, int fd_to, const char *t
         return 0;
 }
 
-static int copy_file_with_version_check(const char *from, const char *to, bool force) {
+static int hashsums_match(int fd_from, const char *from, int fd_to, const char *to) {
+        uint8_t hash_from[SHA256_DIGEST_SIZE], hash_to[SHA256_DIGEST_SIZE];
+        int r;
+
+        assert(fd_from >= 0);
+        assert(from);
+        assert(fd_to >= 0);
+        assert(to);
+
+        r = sha256_fd(fd_from, UINT64_MAX, hash_from);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to compute sha256 for %s: %m", from);
+
+        r = sha256_fd(fd_to, UINT64_MAX, hash_to);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to compute sha256 for %s: %m", to);
+
+        return memcmp(hash_from, hash_to, SHA256_DIGEST_SIZE) == 0;
+}
```

Comment:
why do a cryptographic hash if you just want to compare two files by contents?

just add a proper function compare_fd() or so, which starts by checking that the two inodes refer to regular files, and have the same sizes, and then reads block by block comparing things. That should be a lot more efficient, as you already stop doing the comparison on the first different block you read, and quite often that's already the first one.

Reply:
This is package installation/update path, not critical boot path, so hashing a few hundred KBs worth of data is not worth reinventing a new comparison system when we already have this.. The package upgrade calculates hashes anyway, so it wouldn't make any difference.

Reply:
yeah, still. if you want to compare files, compare files. don't pull in a crypto stack for this. it's ineffecient and pointless.

Reply:
The normal way to compare installed files is with a hashsum. What you propose is premature optimization - show me some benchmarks and profiling that show this one hashsum is the bottleneck when the shim packages get upgraded, and we can consider implementing some custom micro-optimizations. Until then, a hashsum is perfectly adequate for this purpose

Reply:
i am not sure what is normal, but neither "cmp" or "diff" bother the sha256 last time i looked.

writing a proper compare func is not hard, look:

```c
static int data_same_fd(int fda, int fdb) {
        int r;

        struct stat sta, stb; 
        if (fstat(fda, &sta) < 0)
                return -errno;
        r = stat_verify_regular(&sta);
        if (r < 0)
                return r;

        if (fstat(fdb, &stb) < 0)
                return -errno;
        r = stat_verify_regular(&stb);
        if (r < 0)
                return r; 

        if (sta.st_size != stb.st_size)
                return false;

        r = inode_same_at(fda, NULL, fdb, NULL); /* Before we start reading things, let's see if this is the same inode, which would allow us to shortcut things */
        if (r != 0)
                return r;

        for (;;) { 
                uint8_t buffer1[64*1024], buffer2[64*1024];
                 
                ssize_t na, nb;
                na = read(fda, buffer1, sizeof(buffer1));
                if (na < 0)
                        return -errno;

                nb = loop_read(fdb, buffer2, MAX(na, 1));
                if (nb < 0)
                        return nb;

                if (memcmp_nn(buffer1, na, buffer2, nb) != 0)
                        return false;

                if (na == 0)
                        return true;
        }
}
```

Reply:
> i am not sure what is normal, but neither "cmp" or "diff" bother with sha256 last time i looked.

I'm not aware of any package manager using cmp or diff, so that's not really relevant?

> writing a proper compare func is not hard, look:

It's not a question of easy or hard. It's a micro optimization, and like all proposed micro optimizations, it needs to be shown it is worth it. Given this is inherently IO bound, I doubt it makes a meaningful difference, but if you can show that this is indeed the bottleneck when, say, apt-get upgrades the shim packages (shim-helpers-amd64-signed shim-signed shim-signed-common shim-unsigned) then it can be certainly reconsidered

Reply:
sorry, but please let's not be so sloppy, let's do this properly. and this is not going to be io bound at all if you already exit afterthe first different byte.

let's not cut corners all the time please, let's do things properly

Reply:
Exactly - this is doing things properly, by not applying unproven premature optimizations, without any test/bench/profiling that demonstrate they are needed, and instead reusing existing methods

Reply:
dude, i made it so easy for you, wrote you the code, why do we even have to argue about this?

Reply:
Also I regular place 400M kernels as bootx64.efi in my images. It's a concept called USI, unified system image: you stick a whole OS inside a UKI. If bootctl update runs on systems like that (even if that really shouldn't take place) i shouldn't have to wait a minute before it completes.

Reply:
My 5 years old laptop takes ~3s to sha256 a ~600M ISO:

```
$ time sha256sum debian-testing-amd64-netinst.iso 
eceed98dd7238dd1df4d35a300ff4cd916f1d3336bb9d645c4dc3bcd5515a06e  debian-testing-amd64-netinst.iso

real	0m3.756s
user	0m2.881s
sys	0m0.186s
$ ls -lh debian-testing-amd64-netinst.iso 
-rw-r--r-- 1 libvirt-qemu libvirt-qemu 660M Jun 28 22:22 debian-testing-amd64-netinst.iso
```

that is quite far from a minute, 20 times as far to be precise. The fact that you thought it would take 20 times as much really shows this is a premature optimization - the kind for which you usually reject other people's PRs for, like for example https://github.com/systemd/systemd/pull/33514

and that's even an intentional edge case, a normal bootloader that is handled by this would be 600K, not 600M

Reply:
Sorry, but this makes no sense. If we are comparing two files, the cost of reading in both files and calculating the hash is >= the cost of reading in both files and comparing them. The important thing is that if we calculate the hash, we have to read the whole file. But if we're just comparing block by block, we exit on the first block that differs, which means that usually we'll read just a handful of blocks, most likely just one.

Reply:
Of course, but the real question is not that, the real question is: is it worth it or is it negligible.  As I have shown above, the difference appears negligible even on edge cases. Meanwhile 42 new lines of code, at the industry standard rate of one bug every six lines of code, potentially equals to up to 7 new bugs.

Reply:
is this really the hill you want to die on?

Reply:
> Meanwhile 42 new lines of code, at the industry standard rate of one bug every six lines of code, potentially equals to up to 7 new bugs.

Sorry, but in reality barely any new logic will be introduced - it only requires _trivial_ adaptation of `conservasive_renameat`.

Please, as stated, let's do things the right way. There's nothing stopping you from that as the logic is in-place already.